include/maliput/math/kd_tree.h

```diff
@@ -130,45 +231,49 @@ struct NodeCmp {
 ///
 /// Inspired on https://rosettacode.org/wiki/K-d_tree.
 ///
+/// @tparam CRTP Derived class. It follows the curiously recurring template pattern.
 /// @tparam Coordinate Data type being used, must have:
 /// - operator[] for accessing the value in each dimension.
 /// @tparam Dimension Dimension of the KD-tree.
+/// @tparam Region The type of the region.
 /// @tparam Distance A functor used for getting the distance between two coordinates. By default,
 /// details::SquaredDistance is used.
 /// @tparam NodeCmp A functor used for comparing two nodes at certain index/dimension. By default, details::NodeCmp is
 /// used.
-template <typename Coordinate, std::size_t Dimension,
+template <typename CRTP, typename Coordinate, std::size_t Dimension,
+          typename Region = BoundingRegion<Coordinate>,
           typename Distance = details::SquaredDistance<Coordinate, Dimension>,
           typename NodeCmp = details::NodeCmp<Dimension>>
-class KDTree {
+class KDTreeBase {
```

Comment:
I created a KDTree3D that provides a RangeSearch query. Providing a RangeSearch for any n-dimensional kdtree seems not to be trivial therefore so I narrow it down to 3-d in its respective class.

I went for the CRTP pattern:
- As this class is meant to be used in a performant manner I wanted to avoid having references because having polymorphism with virtual methods (vtable references). 
  - It will only affect it if it is used extensible though, so I guess it isn't a strong reason at the end.